### PR TITLE
implement builtin string function left_binary and right_binary

### PR DESCRIPTION
**Pull request by niedhui**
## What have you changed? (mandatory)

1. implement builtin string function `left_binary` and `right_binary`
2. fix the existed `left` and `right` function's length overflow issue

## What are the type of the changes? (mandatory)

Improvement

## How has this PR been tested? (mandatory)

Unit Test

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Refer to a related PR or issue link (optional)

#3275

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 1]
```diff
@@ -175,14 +175,13 @@ impl ScalarFunc {
     ) -> Result<Option<Cow<'a, [u8]>>> {
         let s = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
         let i = try_opt!(self.children[1].eval_int(ctx, row));
-        if i <= 0 {
+        let (i, length_positive) = i64_to_usize(i, self.children[1].is_unsigned());
+        if !length_positive || i == 0 {
             return Ok(Some(Cow::Owned(b"".to_vec())));
         }
-        if s.chars().count() > i as usize {
+        if s.chars().count() > i {
```
[the remark this answers is not in the document]
Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_string.rs]: no, Iter only had `count`, and need to *consume* the Iter

File: src/coprocessor/dag/expr/builtin_string.rs [hunk 2]
```diff
@@ -175,14 +175,13 @@ impl ScalarFunc {
...
             let t = s.chars();
```
[the remark this answers is not in the document]
Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_string.rs]: As `count`  consumes the `Iter`. so we need to create it twice here.

another option is to ignore the count,  even the `count` is larger than i, we still get the right result, but will do some allocation, will rust do some optimization to void this?

File: src/coprocessor/dag/expr/builtin_string.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ice1000) [reply src/coprocessor/dag/expr/builtin_string.rs]: What I mean is `s.len()`, not `s.chars().len()`.
https://doc.rust-lang.org/std/primitive.str.html#method.len

[the remark this answers is not in the document]
Comment (niedhui) [reply src/coprocessor/dag/expr/builtin_string.rs]: this is in the `left` function, not `left_binary`,  we need chars length, not bytes length.

[the remark this answers is not in the document]
Comment (ice1000) [reply src/coprocessor/dag/expr/builtin_string.rs]: Oh! That's very reasonable! Thanks!

**Comment by huachaohuang** [comment]
CI failed but it's not this PR's fault, merge now.

**Pull request merged**